src/seraphis_wallet/transaction_history.h

```diff
@@ -0,0 +1,195 @@
+
+#pragma once
+
+// local headers
+#include "crypto/crypto.h"
+#include "ringct/rctTypes.h"
+#include "seraphis_core/jamtis_destination.h"
+#include "seraphis_core/jamtis_payment_proposal.h"
+#include "seraphis_impl/enote_store.h"
+#include "seraphis_main/contextual_enote_record_types.h"
+#include "seraphis_main/txtype_squashed_v1.h"
+
+// third party headers
+#include <boost/range.hpp>
+#include "boost/range/iterator_range.hpp"
+
+// standard headers
+#include <cstddef>
+#include <cstdint>
+#include <functional>
+#include <map>
+#include <tuple>
+#include <unordered_map>
+#include <unordered_set>
+#include <vector>
+
+// forward declarations
+
+using namespace sp::jamtis;
+using namespace sp;
+
+// struct to store the relevant info from a tx
+struct TransactionRecordV1
+{
+    // key images of spent enotes for tracking purposes
+    std::vector<crypto::key_image> legacy_spent_enotes;
+    std::vector<crypto::key_image> sp_spent_enotes;
+
+    // JamtisPaymentProposals
+    std::vector<JamtisPaymentProposalSelfSendV1> selfsend_payments;
+    std::vector<JamtisPaymentProposalV1> normal_payments;
+
+    // fees and total sent
+    // useful to store here also instead of looking directly at the enotes and blockchain
+    rct::xmr_amount amount_sent;
+    rct::xmr_amount fee_sent;
+};
+
+// struct to efficiently store/sort records
+struct SpTransactionStoreV1
+{
+    // quickly find TransactionRecordV1 from txid
+    serializable_unordered_map<rct::key, TransactionRecordV1> tx_records;
+
+    // sort by timestamp
+    std::vector<std::pair<uint64_t, rct::key>> txs_by_timestamp;
+};
+
+//-----------------------------------------------------------------
+/// Operators
+bool operator==(const SpTransactionStoreV1 &a, const SpTransactionStoreV1 &b);
+bool operator==(const TransactionRecordV1 &a, const TransactionRecordV1 &b);
+//-----------------------------------------------------------------
+
+class SpTransactionHistory
+{
+    SpTransactionStoreV1 m_sp_tx_store;
+
+   public:
+    //-----------------------------------------------------------------
+    /// Add entries to SpTransactionStore
+    //-----------------------------------------------------------------
+
+    /**
+     * brief: add entry to tx_records
+     * param: txid - transaction id
+     * param: record - transaction record
+     */
+    void add_entry_to_tx_records(const rct::key &txid, const TransactionRecordV1 &record);
+
+    /**
+     * brief: add entry to txs_by_timestamp
+     * param: timestamp - timestamp when transaction is created
+     * param: txid - transaction id
+     */
+    void add_entry_to_txs_by_timestamp(const uint64_t timestamp, const rct::key &txid);
+
+    /**
+     * brief: add entries to SpTransactionStore from single_tx
+     * param: single_tx - transaction
+     * param: selfsend_payments - selfsend payments
+     * param: normal_payments - normal payments
+     */
+    void add_single_tx_to_tx_history(const SpTxSquashedV1 &single_tx,
```

Comment:
I think the name part `to_tx_history` is superfluous: We are here in the tx history, to what else would you want to add it? Plus, other methods do not have it either.